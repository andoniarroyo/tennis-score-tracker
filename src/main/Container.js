// @flow
// external modules
import { connect } from 'react-redux';
// internal types
import type { StateType, MatchScoreType } from './index';
// internal modules
import scoreSelector from './selectors';
import TenisScoreTracker from './TennisScoreTracker';
import updateScore from './actions';

type StateToPropsType = {
    score: MatchScoreType
}

/**
 * Maps the application state to props passed to the component
 * @param {StateType} state The application redux state
 * @returns {StateToPropsType} The props coming from the state
 * @private
 */
const mapStateToProps = (state: StateType): StateToPropsType => ({
  score: scoreSelector(state),
});

const mapDispatchToProps = {
  updateScore,
};

export default connect(mapStateToProps, mapDispatchToProps)(TenisScoreTracker);
