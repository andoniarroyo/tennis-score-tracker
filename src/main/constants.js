// @flow
export const SETS_TO_WIN = 3;
export const MATCH_STATUS_INPLAY = 'inplay';
export const MATCH_STATUS_FINISHED = 'finished';
export const PLAYER1_NAME = 'player1';
export const PLAYER2_NAME = 'player2';

// actions related constants
export const UPDATE_SCORE_REQUESTED = 'UPDATE_SCORE_REQUESTED';
