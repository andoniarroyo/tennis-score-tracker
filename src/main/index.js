// @flow
import scoreReducer from './reducers';

export type PlayerScoreType = {
    sets: number,
    games: number,
    score: string,
    tieBreak: number,
    historicScoreboard: Array<number>,
}

export type MatchScoreType = {
    status: string,
    winner: string,
    player1: PlayerScoreType,
    player2: PlayerScoreType,
    inTiebreakMode: boolean,
}

export type StateType = {
    score: MatchScoreType
}

const reducers = {
  score: scoreReducer,
};

export default reducers;
