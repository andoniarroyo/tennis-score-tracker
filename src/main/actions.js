// @flow
// internal types
import type { MatchScoreType } from './index';
// internal modules
import { UPDATE_SCORE_REQUESTED } from './constants';

export type UpdateScoreActionType = {
    type: string,
    newScore: MatchScoreType
};

/**
 * Action creator to update the match score
 * @param {MatchScoreType} newScore The new score value
 * @returns {UpdateScoreActionType} The generated action
 * @public
 */
const updateScore = (newScore: MatchScoreType): UpdateScoreActionType => (
  {
    type: UPDATE_SCORE_REQUESTED,
    newScore,
  }
);

export default updateScore;
