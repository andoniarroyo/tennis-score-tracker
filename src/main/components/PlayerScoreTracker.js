// @flow
// external modules
import React from 'react';
// internal types
import type { PlayerScoreType } from '../index';
// internal modules
import { MATCH_STATUS_FINISHED } from '../constants';

import '../styles/components/playerScoreTracker.scss';

type PlayerScoreTrackerPropsType = {
  name: string,
  status: string,
  playerScore: PlayerScoreType,
  updateScore: Function,
  playerScored: Function
}

/**
 * The tennis score tracker render method
 * @param {TennisScoreTrackerPropsType} The properties to render the object
 * @returns {React$Element<any>} HTML markup for the component
 */
const PlayerScoreTracker = ({
  name, status, playerScore, updateScore, playerScored,
}: PlayerScoreTrackerPropsType): React.Element<*> => (
  <div>
    <div className="player-name">{name}</div>
    <div className="player-score-content">
      <div>
        sets:
      </div>
      <div>
        {playerScore.sets}
      </div>
      <div>
        games:
      </div>
      <div>
        {playerScore.games}
      </div>
      <div>
        score:
      </div>
      <div className="score-main-value">
        {playerScore.score}
      </div>
      <div className="scored">
        <button disabled={status === MATCH_STATUS_FINISHED} onClick={() => updateScore(playerScored())}>
          {name} scored!
        </button>
      </div>
    </div>
  </div>
);

export default PlayerScoreTracker;
