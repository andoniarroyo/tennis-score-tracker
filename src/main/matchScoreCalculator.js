// @flow
// internal types
import type { MatchScoreType } from './index';

// internal modules
import { PLAYER1_NAME, PLAYER2_NAME } from './constants';
import calculateNextPoint from './scoreCalculator/points';
import calculateNextGame from './scoreCalculator/games';
import calculateNextSet from './scoreCalculator/sets';
import calculateNextTieBreak from './scoreCalculator/tieBreak';

export type ScoreContextType = {
    winnerName: string,
    loserName: string,
    matchScore: MatchScoreType,
}
export type UpdateSideEffectType = {
    keyPath: string,
    value: mixed
}
export type UpdateSideEffectsType = Array<UpdateSideEffectType>;

/**
 * Gets the point loser player name
 * @param {string} winnerPlayer Name of the payer who won the point
 * @returns {string} The name of the loser player
 * @private
 */
const getLoserPlayer = (winnerPlayer: string): string => (
  winnerPlayer === PLAYER1_NAME ? PLAYER2_NAME : PLAYER1_NAME
);

/**
 * Clones the current match story
 * @param {MatchScoreType} previousMatchScore The match score object to be cloned
 * @returns {MatchScoreType} The new cloned score
 * @private
 */
const cloneMatchScore = (previousMatchScore: MatchScoreType): MatchScoreType => (
  Object.assign({}, previousMatchScore)
);

/**
 * Applies a side effect to the match score updating the state
 * @param {MatchScoreType} matchScore The match score to be used as reference
 * @param {UpdateSideEffectsType} updatesSideEffects The side effects to be applied
 * @returns {MatchScoreType} The match scrore resulting status
 * @private
 */
const applySideEffect = (matchScore: MatchScoreType, updatesSideEffects: UpdateSideEffectsType): MatchScoreType => {
  const newMatchScore = cloneMatchScore(matchScore);
  updatesSideEffects.forEach((update: UpdateSideEffectType) => {
    const keys = update.keyPath.split(',');
    if (keys.length === 1) {
      newMatchScore[keys[0]] = update.value;
      return newMatchScore;
    }
    let reference = matchScore;
    while (keys.length > 1) {
      reference = reference[keys.shift()];
    }
    reference[keys.shift()] = update.value;
    return reference;
  });
  return newMatchScore;
};

/**
 * Updates the incoming match score based on the player who won the point
 * @param {string} winnerName The name of the point winner
 * @param {MatchScoreType} previousMatchScore The match score to be used as reference
 * @returns {MatchScoreType} The new match score generated
 * @private
 */
const playerScored = (winnerName: string, previousMatchScore: MatchScoreType): MatchScoreType => {
  let matchScore = cloneMatchScore(previousMatchScore);
  const loserName = getLoserPlayer(winnerName);
  const scoreContext = { winnerName, loserName, matchScore };
  // Tiebreak mode has their special flow
  if (previousMatchScore.inTiebreakMode) {
    const newTiebreakSideEffect = calculateNextTieBreak(scoreContext);
    return applySideEffect(matchScore, newTiebreakSideEffect);
  }
  // score
  const newScoreSideEffect = calculateNextPoint(scoreContext);
  matchScore = applySideEffect(matchScore, newScoreSideEffect.updates);
  if (newScoreSideEffect.isFinishingGame) {
    // games
    const newGamesSideEffect = calculateNextGame(scoreContext);
    matchScore = applySideEffect(matchScore, newGamesSideEffect.updates);
    if (newGamesSideEffect.isFinishingSet) {
      // sets
      const newSetsSideEffect = calculateNextSet(scoreContext);
      matchScore = applySideEffect(matchScore, newSetsSideEffect);
    }
  }
  return matchScore;
};

/**
 * Updates the match score based on the player1 scored a point
 * @param {MatchScoreType} previousScore The previous match score
 * @returns {MatchScoreType} The new match score applying the side effects of the players 1 point
 * @public
 */
export const player1Scored = (previousScore: MatchScoreType): MatchScoreType =>
  playerScored(PLAYER1_NAME, previousScore);

/**
 * Updates the match score based on the player2 scored a point
 * @param {MatchScoreType} previousScore The previous match score
 * @returns {MatchScoreType} The new match score applying the side effects of the players 2 point
 * @public
 */
export const player2Scored = (previousScore: MatchScoreType): MatchScoreType =>
  playerScored(PLAYER2_NAME, previousScore);
