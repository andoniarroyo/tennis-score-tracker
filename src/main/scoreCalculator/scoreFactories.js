// @flow
// internal types
import type { PlayerScoreType, MatchScoreType } from '../index';
//internal modules
import { MATCH_STATUS_INPLAY } from '../constants';

/**
 * Composes a playscore with the values received by parameters
 * @param {number} sets The number of sets won by the player
 * @param {number} games The number of games won by the player
 * @param {string} score The number of points won by the player
 * @param {number} tieBreak The number of tiebreak points won by the player
 * @param {Array<number>} historicScoreboard The historic scoreboards for the player
 * @returns {PlayerScoreType} The player score generated object
 */
export const playerScoreFactory = (sets: number = 0, games: number = 0, score: string = '0',
  tieBreak: number = 0, historicScoreboard: Array<number> = []): PlayerScoreType => (
  {
    sets, games, score, tieBreak, historicScoreboard,
  }
);

/**
 * Composes a match score based on the received parameters
 * @param {PlayerScoreType} player1Score The player 1 score
 * @param {PlayerScoreType} player2Score The player 2 score
 * @param {boolean} inTiebreakMode The initial value for the in tiebreak mode
 * @returns {MatchScoreType} The match score generated object
 */
export const matchScoreFactory = (player1Score: PlayerScoreType = playerScoreFactory(),
  player2Score: PlayerScoreType = playerScoreFactory(), inTiebreakMode: boolean = false): MatchScoreType => (
  {
    status: MATCH_STATUS_INPLAY,
    winner: '',
    player1: player1Score,
    player2: player2Score,
    inTiebreakMode,
  }
);
