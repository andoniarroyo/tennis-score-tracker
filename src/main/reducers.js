// @flow
// internal types
import type { UpdateScoreActionType } from './actions';
import type { MatchScoreType } from './index';
// internal modules
import { matchScoreFactory } from './scoreCalculator/scoreFactories';
import { UPDATE_SCORE_REQUESTED } from './constants';

type CommonActionType = {
    type: string,
}

const reducers = {
  [UPDATE_SCORE_REQUESTED]:
        (score: MatchScoreType, { newScore }: UpdateScoreActionType) => newScore,
};

/**
 * Reduces the score in the state based on the received action
 * @param {object} score The current score
 * @param { CommonActionType } action The action to be evaluated
 * @returns {object} The new reduced score
 */
const scoreReducer = (score: MatchScoreType = matchScoreFactory(), action: CommonActionType) => {
  const reducer = reducers[action.type];
  return reducer ? reducer(score, action) : score;
};

export default scoreReducer;
