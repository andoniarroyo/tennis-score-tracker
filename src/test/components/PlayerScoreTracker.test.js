import React from 'react';
import renderer from 'react-test-renderer';

import { MATCH_STATUS_FINISHED } from '../../main/constants';
import PlayerScoreTracker from '../../main/components/PlayerScoreTracker';
import { playerScoreFactory } from '../../main/scoreCalculator/scoreFactories';

describe('<PlayerScoreTracker />', () => {
  it('renders the player score tracker', () => {
    const props = {
      name: 'andoni',
      status: MATCH_STATUS_FINISHED,
      playerScore: playerScoreFactory(),
      updateScore: () => {},
      playerScored: () => {},
    };

    const tree = renderer.create(<PlayerScoreTracker {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
