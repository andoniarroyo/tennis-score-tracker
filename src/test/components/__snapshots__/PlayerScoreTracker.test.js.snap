// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayerScoreTracker /> renders the player score tracker 1`] = `
<div>
  <div
    className="player-name"
  >
    andoni
  </div>
  <div
    className="player-score-content"
  >
    <div>
      sets:
    </div>
    <div>
      0
    </div>
    <div>
      games:
    </div>
    <div>
      0
    </div>
    <div>
      score:
    </div>
    <div
      className="score-main-value"
    >
      0
    </div>
    <div
      className="scored"
    >
      <button
        disabled={true}
        onClick={[Function]}
      >
        andoni
         scored!
      </button>
    </div>
  </div>
</div>
`;
